feat: create @lavamoat/policy

boneskull opened the pull request:
This is an experimental PR (for funsies) that uses [Zod v4](https://zod.dev) to generate both a LavaMoat policy validator and a JSON schema.

The schema and policy generation live in a new package, `@lavamoat/policy`.

The sources are TS.  I attempted to use native Node.js type stripping for running tests, but a) this limits the Node.js versions available to the test matrix, and b) it does not work with _type-only_ packages (of which `@lavamoat/types` is one; `type-fest` is another we often use).  I pulled in [tsx](https://npm.im/tsx) to solve this, but I could be persuaded to convert the tests to JS.  **The tests are written using the native `node:test` runner; not AVA.**

`@lavamoat/policy` should _probably_ be a CommonJS module, but it is not.  AFAICT, CJS is incompatible with type stripping, and I didn't revert the target module system after I had to abandon it

It supersedes #733.

boneskull commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/LavaMoat/LavaMoat/1670?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#1670** <a href="https://app.graphite.dev/github/pr/LavaMoat/LavaMoat/1670?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/LavaMoat/LavaMoat/1670?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#1593** <a href="https://app.graphite.dev/github/pr/LavaMoat/LavaMoat/1593?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#1620** <a href="https://app.graphite.dev/github/pr/LavaMoat/LavaMoat/1620?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>: 1 other dependent PR ([#1547](https://github.com/LavaMoat/LavaMoat/pull/1547) <a href="https://app.graphite.dev/github/pr/LavaMoat/LavaMoat/1547?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>)
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

naugtur commented:
This is neat. I'd feel safer if we had a bunch of tests that make sure existing policy examples pass the validation so that we don't make something suddenly invalid, but other than that I want it merged

boneskull commented:
Note: as written, the JSON schema will validate a schema using invalid (per regex) canonical names, due to a limitation in Zod where a `ZodRegex` cannot be a key of a `ZodRecord`.  There may be a workaround

boneskull commented:
I've added tests for this, but I still think it belongs in its own package.

boneskull committed:
feat(policy): create @lavamoat/policy

This creates `@lavamoat/policy`, a package which centralizes the definition of a LavaMoat policy, provides validation, and contains a JSON schema.

boneskull committed:
feat(policy): create @lavamoat/policy

This creates `@lavamoat/policy`, a package which centralizes the definition of a LavaMoat policy, provides validation, and contains a JSON schema.

boneskull commented:
<!-- gh-stack:nav -->
### :books: Pull Request Stack

- branch: `main`
  - **[feat: create @lavamoat/policy #1670](https://github.com/LavaMoat/LavaMoat/pull/1670)** - branch: `boneskull/experimental-zod` *(this PR)*

---
*Managed by [gh-stack](https://github.com/boneskull/gh-stack)*
